pluto: Add hyper-proxy as a module

jpculp opened the pull request:
**Description of changes:**

The hyper_proxy module is a clone of [tafia/hyper-proxy](https://github.com/tafia/hyper-proxy), but modified to take advantage of newer Rust dependencies.
Since Bottlerocket only uses 'rustls', other features and unused lines of code have been removed.

Also bumps `webpki` to **0.22.1**.

**Testing done:**

Launched instance using an `aws-k8s-1.25` AMI with user-data that includes the following:

```toml
[settings.network]
https-proxy = "<my-proxy>:9898"
```

Checked proxy server logs and found all traffic from my private IP being tunneled, including the EKS API calls for `DescribeCluster`:

```
1694211359.018    163 192.168.81.185 TCP_TUNNEL/200 9071 CONNECT eks.us-west-2.amazonaws.com:443 - HIER_DIRECT/35.164.66.95 -
1694211359.263    147 192.168.81.185 TCP_TUNNEL/200 9071 CONNECT eks.us-west-2.amazonaws.com:443 - HIER_DIRECT/35.164.66.95 -
1694211361.058    340 192.168.81.185 TCP_TUNNEL/200 20319 CONNECT api.ecr.us-west-2.amazonaws.com:443 - HIER_DIRECT/34.223.26.183 -
1694211361.070     77 192.168.81.185 TCP_TUNNEL/200 9528 CONNECT api.ecr.us-west-2.amazonaws.com:443 - HIER_DIRECT/34.223.26.183 -
1694211361.070     49 192.168.81.185 TCP_TUNNEL/200 7644 CONNECT api.ecr.us-west-2.amazonaws.com:443 - HIER_DIRECT/34.223.26.183 -
```

Checking CloudTrail, I can see that the `DescribeCluster` API call source IP is my proxy server:

```
    "eventTime": "2023-09-08T22:15:59Z",
    "eventSource": "eks.amazonaws.com",
    "eventName": "DescribeCluster",
    "awsRegion": "us-west-2",
    "sourceIPAddress": <my-proxy>,
    "userAgent": "aws-sdk-rust/0.55.3 os/linux lang/rust/1.71.1",
```

I set no-proxy for the EKS API endpoint and reboot like so:

```bash
apiclient set --json '{"settings":{"network":{"no-proxy":["eks.us-west-2.amazonaws.com"]}}}'
apiclient reboot
```

Then I can see that EKS API calls do not get proxied and the source IP is my actual instance's public IP:

```
    "eventTime": "2023-09-08T22:22:05Z",
    "eventSource": "eks.amazonaws.com",
    "eventName": "DescribeCluster",
    "awsRegion": "us-west-2",
    "sourceIPAddress": "54.212.110.209",
    "userAgent": "aws-sdk-rust/0.55.3 os/linux lang/rust/1.71.1",
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp commented:
I've included a temp commit to make it easier to see what changed in the `hyper-proxy` code, which I plan to squash prior to merge.

jpculp commented:
In its current state `--no-default-features` or setting the feature to `rustls-base` (which you shouldn't), results in compilation errors. Since we only ever plan on using `rustls`, we can simply remove the conditional compilation blocks. Does anyone have any opinions on that?

jpculp commented:
Removed extraneous _(to Bottlerocket)_ features and re-tested.

bcressey commented on sources/deny.toml:
```diff
@@ -25,6 +25,7 @@ allow = [
 
 exceptions = [
     { name = "unicode-ident", version = "1.0.4", allow = ["MIT", "Apache-2.0", "Unicode-DFS-2016"] },
+    { name = "webpki-roots", allow = ["MPL-2.0"], version = "*" },
```
We shouldn't allow `webpki-roots` to come in, as it means that something might not be using the CA certificates we manage.

jpculp replied:
Whoops. I meant to remove this in my last push, but missed it. It was being pulled in to support the `rustls-webpki` feature, but with that feature removed it won't get pulled in.

bcressey commented on sources/hyper-proxy-shim/README.md:
```diff
@@ -0,0 +1,71 @@
+# hyper-proxy
+
+[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE)
+[![crates.io](http://meritbadge.herokuapp.com/hyper-proxy)](https://crates.io/crates/hyper-proxy)
+
+A proxy connector for [hyper][1] based applications.
```
If we're forking this code, I would expect a shorter README.md pointing to the original source, by expanding the credits section. We can probably drop the example.

bcressey commented on sources/hyper-proxy-shim/src/stream.rs:
```diff
@@ -0,0 +1,91 @@
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
```
Need to figure out what license header to put in these files.

I'd expect something along the lines of what we do for the [cargo macros](https://github.com/bottlerocket-os/bottlerocket-sdk/blob/6d7ae094de951a92ea564a3c7ccd164a858d19b4/macros/cargo#L4).

bcressey commented on sources/Cargo.toml:
```diff
@@ -103,8 +103,14 @@ members = [
 [profile.release]
 debug = true
+
+# This replaces hyper-proxy with a shim using updated dependencies
+[patch.crates-io.hyper-proxy]
+path = "hyper-proxy-shim"
```
Do we need to call it `hyper-proxy-shim` or can it just be `hyper-proxy`?

bcressey commented on sources/hyper-proxy-shim/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+keywords = ["hyper", "proxy", "tokio", "ssl"]
+categories = ["web-programming::http-client", "asynchronous", "authentication"]
+license = "MIT"
+edition = "2021"
```
```suggestion
edition = "2021"
publish = false
```

bcressey commented on sources/hyper-proxy-shim/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+name = "hyper-proxy"
+version = "0.10.0"
+authors = ["Johann Tuffe <tafia973@gmail.com>"]
+description = "A proxy connector for Hyper-based applications"
```
We need guidance on this too, but we might want to leave the original author and add an author below, with a comment that 99% of the work is Johann's and the rest was just dropping unused features and updating dependencies.

webern commented on tools/Cargo.lock:
```diff
@@ -3769,9 +3769,9 @@ dependencies = [
 [[package]]
 name = "webpki"
-version = "0.22.0"
+version = "0.22.1"
```
How did this end up affecting tools? The tools Cargo.lock is now gone.

jpculp replied:
It was a dependency of `pubsys` coming from `coldsnap` and `aws-smithy-client`.

jpculp committed:
pluto: add hyper-proxy as a module

The 'hyper_proxy' module is clone of tafia/hyper-proxy, but modified to
take advantage of newer Rust dependencies. Since Bottlerocket only uses
'rustls', other features and unused lines of code have been removed.

jpculp commented:
Rather than adding `hyper-proxy` to `sources`, we'll just add it as a module for `pluto`.

webern commented on sources/api/pluto/src/proxy.rs:
```diff
@@ -1,5 +1,6 @@
+use crate::hyper_proxy::{Proxy, ProxyConnector};
+
 use hyper::Uri;
```
nit: I prefer to keep these together and let rustfmt sort it. Not sure if we have an established style on this though.

```suggestion
use crate::hyper_proxy::{Proxy, ProxyConnector};
use hyper::Uri;
```

webern reviewed (approved):
Makes sense to me.

jpculp commented:
Re-based, re-tested, and ready for review.

bcressey commented on COPYRIGHT:
```diff
@@ -17,3 +17,7 @@ operating system images.
 macros/rust and macros/cargo (used during build) are derived from the Fedora Rust SIG's rust2rpm.
 https://pagure.io/fedora-rust/rust2rpm
 Copyright (c) 2017 Igor Gnatenko
```
Can you add a separate commit that removes this? It's no longer applicable here; we need to add it to COPYRIGHT in the sdk repo.

bcressey commented on sources/api/pluto/Cargo.toml:
```diff
@@ -11,9 +11,12 @@ exclude = ["README.md"]
 
 [dependencies]
 apiclient = { path = "../apiclient", version = "0.1" }
+bytes = "1"
```
I believe the license here should now be:
```
license = (APACHE-2.0 OR MIT) AND MIT
```

jpculp committed:
copyright: remove macros/rust and macros/cargo
